shell/browser/win/dark_mode.cc

```diff
@@ -0,0 +1,195 @@
+// Copyright (c) 2020 Microsoft Inc. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE-CHROMIUM file.
+
+#include "shell/browser/win/dark_mode.h"
+
+#include <dwmapi.h>  // DwmSetWindowAttribute()
+
+#include "base/files/file_path.h"
+#include "base/scoped_native_library.h"
+#include "base/win/pe_image.h"
+#include "base/win/win_util.h"
+
+// This namespace contains code originally from
+// https://github.com/ysc3839/win32-darkmode/
+// governed by the MIT license and (c) Richard Yu
+namespace {
+
+// 1903 18362
+enum PreferredAppMode { Default, AllowDark, ForceDark, ForceLight, Max };
+
+bool g_darkModeSupported = false;
+bool g_darkModeEnabled = false;
+DWORD g_buildNumber = 0;
+
+enum WINDOWCOMPOSITIONATTRIB {
+  WCA_USEDARKMODECOLORS = 26  // build 18875+
+};
+struct WINDOWCOMPOSITIONATTRIBDATA {
+  WINDOWCOMPOSITIONATTRIB Attrib;
+  PVOID pvData;
+  SIZE_T cbData;
+};
+
+using fnSetWindowCompositionAttribute =
+    BOOL(WINAPI*)(HWND hWnd, WINDOWCOMPOSITIONATTRIBDATA*);
+fnSetWindowCompositionAttribute _SetWindowCompositionAttribute = nullptr;
+
+bool IsHighContrast() {
+  HIGHCONTRASTW highContrast = {sizeof(highContrast)};
+  if (SystemParametersInfoW(SPI_GETHIGHCONTRAST, sizeof(highContrast),
+                            &highContrast, FALSE))
+    return highContrast.dwFlags & HCF_HIGHCONTRASTON;
+  return false;
+}
+
+void RefreshTitleBarThemeColor(HWND hWnd, bool dark) {
+  LONG ldark = dark;
+  if (g_buildNumber >= 20161) {
+    // DWMA_USE_IMMERSIVE_DARK_MODE = 20
+    DwmSetWindowAttribute(hWnd, 20, &ldark, sizeof dark);
+    return;
+  }
+  if (g_buildNumber >= 18363) {
+    auto data = WINDOWCOMPOSITIONATTRIBDATA{WCA_USEDARKMODECOLORS, &ldark,
+                                            sizeof ldark};
+    _SetWindowCompositionAttribute(hWnd, &data);
+    return;
+  }
+  DwmSetWindowAttribute(hWnd, 0x13, &ldark, sizeof ldark);
+}
+
+bool CheckBuildNumber(DWORD buildNumber) {
+  return (buildNumber == 17763 ||  // 1809
+          buildNumber == 18362 ||  // 1903
+          buildNumber == 18363 ||  // 1909
+          buildNumber == 19041);   // 2004
+}
+
+void InitDarkMode() {
+  // get the "get version & build numbers" function
+  auto nt_snl =
+      base::ScopedNativeLibrary(base::FilePath(FILE_PATH_LITERAL("ntdll.dll")));
+  auto nt_pei = base::win::PEImage(nt_snl.get());
+  using fnRtlGetNtVersionNumbers = VOID(WINAPI*)(LPDWORD, LPDWORD, LPDWORD);
+  auto RtlGetNtVersionNumbers = reinterpret_cast<fnRtlGetNtVersionNumbers>(
+      nt_pei.GetProcAddress("RtlGetNtVersionNumbers"));
+  if (!RtlGetNtVersionNumbers) {
+    return;
+  }
+
+  // check to see if the major/minor/build support these dark mode APIS
+  DWORD major = 0, minor = 0;
+  RtlGetNtVersionNumbers(&major, &minor, &g_buildNumber);
+  g_buildNumber &= ~0xF0000000;
+  if (major != 10 || minor != 0 || !CheckBuildNumber(g_buildNumber)) {
```

Comment:
Can you use `base::win::GetVersion` instead, this seems like you will have to update `CheckBuildNumber` for every OS release.

Reply:
Yes for certain we can use `base::win::OSInfo` instead, that's a great idea.

But... yes, absolutely there will be upkeep on this because there is still no documented API for this and it could all change in a future version of Windows. See for example how ordinal 135 in uxtheme changed from `AllowDarkModeForApp()` to `SetPreferredAppMode()` in version 1903. It's better for dark mode to just stop working in newer versions than it is to potentially call an unknown function signature.

Until there is supported stable API for these calls, I don't see a safe alternative.

Reply:
Migrated build/version checks to `base::win::OSInfo` in b1d76af6d4f386c06e871bbe27e6af3c004b2f68